talk/morelanguage/raii.tex

```diff
@@ -256,6 +256,111 @@
   \end{block}
 \end{frame}
 
+\begin{frame}[fragile]
+  \frametitlecpp[11]{weak\_ptr}
+  \begin{block}{weak\_ptr: a non-owning observer}
+    \small
+    \begin{itemize}
+    \item Sometimes want to observe resources without keeping them alive
+    \begin{itemize}
+      \item \mintinline{cpp}{shared_ptr}? Resource stays alive
+      \item Bare pointer? Risk of dangling pointer when resource is out of scope
+    \end{itemize}
+    \item The solution is to use \mintinline{cpp}{weak_ptr} constructed from a shared pointer
+    \item To access resource, convert weak into a \mintinline{cpp}{shared_ptr}
+    \end{itemize}
+  \end{block}
+  \begin{exampleblock}{}
+    \begin{cppcode*}{}
+      std::weak_ptr<Cache> weakPtr{ getSharedCache() };
+      // ...
+      if (std::shared_ptr<Cache> cache = weakPtr.lock()) {
+        // Cache is alive, we actively extend its lifetime
+        return cache->findItem(...);
+      } else {
+        return recomputeValues(...);
+      }
```

Comment:
It would be awesome to complete the example by the cache update, and maybe show the signature of `getSharedCache`:

```suggestion
      // auto getSharedCache() -> std::weak_ptr<Cache>&;
      auto& weakPtr{ getSharedCache() };
      // ...
      if (std::shared_ptr<Cache> cache = weakPtr.lock()) {
        // Cache is alive, we actively extend its lifetime
        return cache->findItem(...);
      } else {
        return (weakPtr = recomputeCache(...))->findItem(...);
      }
```
You could also consider turning `getSharedCache()` into a static variable (at the cost of potential thread-safety). Please also simplify my suggestion as you like!

Reply:
You interpreted the omission differently than I did. That's a sign that it wasn't complete enough. I tried to be more clear now, please have a look again.